xcmailr-webapp/src/main/java/controllers/MessageListener.java

```diff
@@ -82,6 +89,85 @@ public boolean accept(String from, String recipient)
         return false;
     }
 
+    /**
+     * Checks if forwarding the mail could trigger a loop Checks if the return-path header is empty Checks if the
+     * References/ In-Reply-To Header include a mail from this domain Checks if the custom X-Loop Header exists and
+     * checks it's content The number of hops/ to many Received fields is checked by a internal library
+     * 
+     * @param mail
+     *            the mail to check
+     * @return null if there was no loop, an error message if there was
+     * @throws MessagingException
+     */
+    private String checkForLoop(MimeMessage mail) throws MessagingException
+    {
+        String errorMessage;
+
+        // check the first return-path header
+        String[] returnPathHeaders = mail.getHeader("Return-Path");
+        String returnPathHeader = returnPathHeaders[0];
+        if (returnPathHeader.equals("") || returnPathHeader.equals("<>") || returnPathHeader.equals("< >"))
+        {
+            // loop detected;
+            errorMessage = "Return-Path is empty";
+            return errorMessage;
+        }
+       
+        // check custom X-Loop Header       
+        String customHeader = mail.getHeader(loopHeader, "###");
```

Comment:
In case another email server also added an "X-Loop" Header. Then null would just get the first header and I don't know if that'd be our custom header or theirs.